### DISCUSSION TITLE
Add build flag to enable PGO for Engine build.

### DISCUSSION THREAD

**Discussion by Nobutarou** [OpenRadioss/OpenRadioss#393] [Ideas]
My idea is 
1st, build with '-addflag="-O2 -fprofile-generate''.
2nd, run a target model to collect profile.
3rd, build with '-addflag="-O2 -fprofile-use"'.

However, the 1st build does not succeed. A lot of the following errors happen,

```
/bin/ld: CMakeFiles/engine_linux64_gf_ompi.dir/source/coupling/rad2rad/rad2rad_c.c.o: warning: relocation against `__gcov_time_profiler_counter' in read-only section `.text'
/bin/ld: CMakeFiles/engine_linux64_gf_ompi.dir/aur/OpenRadioss/common_source/comm/compress_nnz.F.o: in function `compress_r_nnz_':
/aur/OpenRadioss/common_source/comm/compress_nnz.F:153: undefined reference to `__gcov_indirect_call'
```

I attach the log file, too.
[engine_build.log](https://github.com/OpenRadioss/OpenRadioss/files/9505268/engine_build.log)

**Comment by Nobutarou** [comment]
I can build.

For the 1st step, I missed -lgcov. Thus, I add it to addflag,

```
-addflag="-O2 -fprofile-generate -lgcov''
```

and I need to pass '-lgcov' to the linker.  I have to modify <openradioss_root>/engine/cbuild_linux64_gf_ompi/CMakeFiles/engine_linux64_gf_ompi.dir/link.txt to

```
gfortran   -fopenmp  -ldl -lrt -lstdc++    @CMakeFiles/engine_linux64_gf_ompi.dir/objects1.rsp @CMakeFiles/engine_linux64_gf_ompi.dir/objects2.rsp -o engine_linux64_gf_ompi  -lrt -L/opt/openmpi/lib -lmpi -lmpi_mpifh -ldl -Wunused-function -lstdc++ -lgcov
```

I don't know what file generates link.txt. Thus, I just modify and disallow removal.

**Comment by laurent-altr** [comment]
Hi @Nobutarou, 

For the engine, you can add the compilation flags into the file [cmake_linux64_gf.txt](https://github.com/OpenRadioss/OpenRadioss/blob/main/engine/CMake_Compilers/cmake_linux64_gf.txt).

Thank you!

**Comment by Nobutarou** [comment]
I find CMAKE_EXE_LINKER_FLAGS in the file. Helpful.
Thank you.

**Comment by zamazan4ik** [comment]
@Nobutarou excuse me for the necroposting. I see you built the project with PGO. Could you please share the benchmark results, how many in performance did you win after enabling PGO for OpenRadioss?

I am interested in this information since I collect all benchmarks about PGO in my PGO repo: https://github.com/zamazan4ik/awesome-pgo

Thank you!

**Comment by Nobutarou** [comment]
Hi,

I don't have the logs but have a memo. According to the memo,

CPU: Ryzen 3 Pro 4350G (4 cores 8 threads)
Ram: 16GB
OS: Arch Linux

I use this input model, and run three times both for normal build and PGO
build to measure their elapsed time,
https://openradioss.atlassian.net/wiki/spaces/OPENRADIOSS/pages/11075585/Bumper+Beam

Here is the result, 15% faster.

| trial |normal build[s] | PGO[s] | speed up |
| ---   | ---                  | ---       | ---         |
|1st    | 374.45           | 323.69 | 1.16 |
|2nd   | 373.87           |  324.11  |1.15 |
| 3rd |         374.43    | 323.42 | 1.16 |

For normal build, I use just "-O2".
For PGO build, I first use "-O2 -fprofile-generate -lgcov" to correct the
profile and re-build with "-O2 -fprofile-use"

Thanks,
Nobuyuki


2024年7月6日(土) 19:16 Alexander Zaitsev ***@***.***>:
